Python/ceval.c

```diff
@@ -3173,6 +3174,25 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
                     la = &co_opcache->u.la;
                     if (la->type == type && la->tp_version_tag == type->tp_version_tag)
                     {
+                        // Hint >= 0 is a dict index; hint < 0 is an inverted slot index
+                        if (la->hint < 0) {
+                            /* Even faster path -- slot hint */
+                            Py_ssize_t offset = ~la->hint;
+                            char *addr = (char *)owner + offset;
+                            res = *(PyObject **)addr;
+                            if (res != NULL) {
+                                Py_INCREF(res);
+                                SET_TOP(res);
+                                Py_DECREF(owner);
+                                DISPATCH();
+                            }
+                            else {
+                                Py_DECREF(owner);
+                                goto error;
```

Comment:
This is going to the error label without setting an exception. Is this an error case or should we just go to the regular path ?

Reply:
We could just raise an AttributeError here, but I'd say falling back and going the regular well tested path is a safer option.

Reply:
It may be easier to just raise than make it so it goes through the right sequence of conditionals without adding moro goto labels. Unless I miss something we just need to:

```
                               SET_TOP(NULL);
                               _PyErr_SetString(tstate, PyExc_AttributeError, .........);
                               goto error;
```

Reply:
Wouldn't that break `__getattr__`?

```
>>> class F:
...   __slots__ = ('a',)
...   def __init__(self):
...    self.a = 1
...   def __getattr__(self, k):
...    print(self, k)
...
>>> f = F()
>>> f.a
1
>>> f.b
<__main__.F object at 0x10bbfb070> b
>>>
```

Reply:
We have this to populate the chache:

```
                if (co_opcache != NULL &&  type->tp_getattro == PyObject_GenericGetAttr)
```

Reply:
Yeah, I think we should follow the already established way and simply fallback to the slow path (calling `PyObject_GetAttr(owner, name);`) if the slot is unused.

Reply:
Wait a second, isn't `__getattr__` *also* `tp_getattro`?:

https://github.com/python/cpython/blob/f4507231e3f0cf8827cec5592571ce371c6813e8/Objects/typeobject.c#L6866-L6876

Reply:
> In any case, I agree that we should try to fall back to the slow path unless is much more complicated.

Yeah, it would be easier to review/read/reason about the code that way.

Reply:
Sorry I didn't see this before. We should do exactly what happens in `PyMember_GetOne()` for `case T_OBJECT_EX:`. It does this:
```
PyErr_SetString(PyExc_AttributeError, l->name);
```
So I'll add that.

Reply:
Oh wait, the name isn't available in the hint. So I'll just fall back to the fast path as Pablo said.